pkg/sql/catalog/tabledesc/constraint.go

```diff
@@ -11,215 +11,477 @@
 package tabledesc
 
 import (
+	"fmt"
+
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
-	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/cockroach/pkg/util"
 )
 
-// constraintCache contains precomputed slices of constraints, categorized by kind and validity.
-// A constraint is considered
-// - active if its validity is VALIDATED;
-// - inactive if its validity is VALIDATING or UNVALIDATED;
-// - dropping if its validity is DROPPING;
-type constraintCache struct {
-	all                  []catalog.Constraint
-	allActiveAndInactive []catalog.Constraint
-	allActive            []catalog.Constraint
+type constraintBase struct {
+	maybeMutation
+}
 
-	allChecks                  []catalog.Constraint
-	allActiveAndInactiveChecks []catalog.Constraint
-	allActiveChecks            []catalog.Constraint
+// AsCheck implements the catalog.ConstraintProvider interface.
```

Comment:
I suggest we always have a line like `var _ catalog.ConstraintProvider = constraintBase{}` to enforce implementation of an interface in compile time.

Reply:
Okay, I see you are doing this for the specific constraint type (like checkConstraint). Great!

Reply:
Thanks but yeah it's a good point. IIRC I try to have the assertions on the leaves of the type hierarchy.